PR: https://github.com/odoo/odoo/pull/203987

From: 6adaab6abb5494fdd0157b502e8f3abc32757420
From: clesgow
Date: 2025-08-29 06:50:52

Breaking data model changes scores: del:24 + add:0, change matches:
-    package_level_id = fields.Many2one('stock.package_level', 'Package Level', check_company=True, copy=False, index='btree_not_null')
-    package_id = fields.Many2one(
-    picking_id = fields.Many2one('stock.picking', 'Picking', check_company=True, index='btree_not_null')
-    move_ids = fields.One2many('stock.move', 'package_level_id')
-    move_line_ids = fields.One2many('stock.move.line', 'package_level_id')
-    location_id = fields.Many2one('stock.location', 'From', compute='_compute_location_id', check_company=True)
-    location_dest_id = fields.Many2one(
-    is_done = fields.Boolean('Done', compute='_compute_is_done', inverse='_set_is_done')
-    state = fields.Selection([
-    is_fresh_package = fields.Boolean(compute='_compute_fresh_pack')
-    picking_type_code = fields.Selection(related='picking_id.picking_type_code')
-    show_lots_m2o = fields.Boolean(compute='_compute_show_lot')
-    show_lots_text = fields.Boolean(compute='_compute_show_lot')
-    company_id = fields.Many2one('res.company', 'Company', required=True, index=True)
-    move_ids_without_package = fields.One2many(
-    move_line_ids_without_package = fields.One2many('stock.move.line', 'picking_id', 'Operations without package', domain=['|',('package_level_id', '=', False), ('picking_type_entire_packs', '=', False)])
-    package_level_ids = fields.One2many('stock.package_level', 'picking_id')
-    package_level_ids_details = fields.One2many('stock.package_level', 'picking_id')

Total Changes: 280

[IMP] stock, *: remove stock.package_level

As packages are now always handled on the move lines, we don't need the
`stock.package_level` model anymore. This means we're no longer
separating moves with/without packages.

Task-4314600

closes odoo/odoo#203987

Related: odoo/enterprise#82453
Related: odoo/upgrade#7576
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock/models/__init__.py
+++ b/addons/stock/models/__init__.py
@@ -22,7 +22,6 @@ from . import stock_scrap
 from . import product
 from . import product_catalog_mixin
 from . import stock_package_history
-from . import stock_package_level
 from . import stock_package_type
 from . import stock_package
 from . import stock_storage_category

--- a/addons/stock/models/stock_move.py
+++ b/addons/stock/models/stock_move.py
@@ -184,7 +184,6 @@ class StockMove(models.Model):
     is_quantity_done_editable = fields.Boolean('Is quantity done editable', compute='_compute_is_quantity_done_editable')
     reference = fields.Char(compute='_compute_reference', string="Reference", store=True)
     move_lines_count = fields.Integer(compute='_compute_move_lines_count')
-    package_level_id = fields.Many2one('stock.package_level', 'Package Level', check_company=True, copy=False, index='btree_not_null')
     display_assign_serial = fields.Boolean(compute='_compute_display_assign_serial')
     display_import_lot = fields.Boolean(compute='_compute_display_assign_serial')
     next_serial = fields.Char('First SN/Lot')
@@ -1163,8 +1162,7 @@ Please change the quantity done or the rounding precision in your settings.""",
         fields = [
             'product_id', 'price_unit', 'procure_method', 'location_id', 'location_dest_id', 'location_final_id',
             'product_uom', 'restrict_partner_id', 'scrapped', 'origin_returned_move_id',
-            'package_level_id', 'propagate_cancel', 'description_picking',
-            'never_product_template_attribute_value_ids',
+            'propagate_cancel', 'description_picking', 'never_product_template_attribute_value_ids',
         ]
         if self.env['ir.config_parameter'].sudo().get_param('stock.merge_only_same_date'):
             fields.append('date')
@@ -1954,8 +1952,7 @@ Please change the quantity done or the rounding precision in your settings.""",
                         assigned_moves_ids.add(move.id)
                         continue
                     # Reserve new quants and create move lines accordingly.
-                    forced_package_id = move.package_level_id.package_id or None
-                    taken_quantity = move._update_reserved_quantity(need, move.location_id, package_id=forced_package_id, strict=False)
+                    taken_quantity = move._update_reserved_quantity(need, move.location_id, strict=False)
                     if float_is_zero(taken_quantity, precision_rounding=rounding):
                         continue
                     moves_to_redirect.add(move.id)

--- a/addons/stock/models/stock_move_line.py
+++ b/addons/stock/models/stock_move_line.py
@@ -263,7 +263,7 @@ class StockMoveLine(models.Model):
             excluded_smls = set(smls.ids)
             if package.package_type_id:
                 best_loc = smls.move_id.location_dest_id.with_context(exclude_sml_ids=excluded_smls, products=smls.product_id)._get_putaway_strategy(self.env['product.product'], package=package)
-                smls.location_dest_id = smls.package_level_id.location_dest_id = best_loc
+                smls.location_dest_id = best_loc
             elif package:
                 used_locations = set()
                 for sml in smls:
@@ -275,8 +275,6 @@ class StockMoveLine(models.Model):
                 if len(used_locations) > 1:
                     for move, grouped_smls in smls.grouped('move_id').items():
                         grouped_smls.location_dest_id = move.location_dest_id
-                else:
-                    smls.package_level_id.location_dest_id = smls.location_dest_id
             else:
                 for sml in smls:
                     putaway_loc_id = sml.move_id.location_dest_id.with_context(exclude_sml_ids=excluded_smls)._get_putaway_strategy(
@@ -450,17 +448,6 @@ class StockMoveLine(models.Model):
             if key in vals:
                 updates[key] = vals[key] if isinstance(vals[key], models.BaseModel) else self.env[model].browse(vals[key])
 
-        if 'result_package_id' in updates:
-            for ml in self.filtered(lambda ml: ml.package_level_id):
-                if updates.get('result_package_id'):
-                    ml.package_level_id.package_id = updates.get('result_package_id')
-                else:
-                    # TODO: make package levels less of a pain and fix this
-                    package_level = ml.package_level_id
-                    ml.package_level_id = False
-                    # Only need to unlink the package level if it's empty. Otherwise will unlink it to still valid move lines.
-                    if not package_level.move_line_ids:
-                        package_level.unlink()
         # When we try to write on a reserved move line any fields from `triggers`, result_package_id excepted,
         # or directly reserved_uom_qty` (the actual reserved quantity), we need to make sure the associated
         # quants are correctly updated in order to not make them out of sync (i.e. the sum of the
@@ -577,12 +564,8 @@ class StockMoveLine(models.Model):
             if not float_is_zero(ml.quantity_product_uom, precision_digits=precision) and ml.move_id and not ml.move_id._should_bypass_reservation(ml.location_id):
                 self.env['stock.quant']._update_reserved_quantity(ml.product_id, ml.location_id, -ml.quantity_product_uom, lot_id=ml.lot_id, package_id=ml.package_id, owner_id=ml.owner_id, strict=True)
         moves = self.mapped('move_id')
-        package_levels = self.package_level_id
         packages = self.env['stock.package'].browse(self.result_package_id._get_all_package_dest_ids())
         res = super().unlink()
-        package_levels = package_levels.filtered(lambda pl: not (pl.move_line_ids or pl.move_ids))
-        if package_levels:
-            package_levels.unlink()
         if moves:
             # Add with_prefetch() to set the _prefecht_ids = _ids
             # because _prefecht_ids generator look lazily on the cache of move_id
@@ -1005,7 +988,6 @@ class StockMoveLine(models.Model):
             'restrict_partner_id': self.picking_id.owner_id.id,
             'company_id': self.picking_id.company_id.id,
             'partner_id': self.picking_id.partner_id.id,
-            'package_level_id': self.package_level_id.id,
         }
 
     def _copy_quant_info(self, vals):
@@ -1087,15 +1069,6 @@ class StockMoveLine(models.Model):
                 package=package
             )
         self.write({'result_package_id': package.id})
-        if len(self.picking_id) == 1:
-            self.env['stock.package_level'].with_context(from_put_in_pack=True).create({
-                'package_id': package.id,
-                'picking_id': self.picking_id.id,
-                'location_id': False,
-                'location_dest_id': self.location_dest_id.id,
-                'move_line_ids': [Command.set(self.ids)],
-                'company_id': self.company_id.id,
-            })
         return package
 
     def _post_put_in_pack_hook(self, package):

--- a/addons/stock/models/stock_package_level.py
+++ b/None
@@ -1,212 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from collections import defaultdict
-
-from odoo import _, api, fields, models
-from odoo.tools.float_utils import float_is_zero
-
-
-class StockPackage_Level(models.Model):
-    _name = 'stock.package_level'
-    _description = 'Stock Package Level'
-    _check_company_auto = True
-
-    package_id = fields.Many2one(
-        'stock.package', 'Package', required=True, check_company=True,
-        domain="[('location_id', 'child_of', parent.location_id), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
-    picking_id = fields.Many2one('stock.picking', 'Picking', check_company=True, index='btree_not_null')
-    move_ids = fields.One2many('stock.move', 'package_level_id')
-    move_line_ids = fields.One2many('stock.move.line', 'package_level_id')
-    location_id = fields.Many2one('stock.location', 'From', compute='_compute_location_id', check_company=True)
-    location_dest_id = fields.Many2one(
-        'stock.location', 'To', check_company=True,
-        compute="_compute_location_dest_id", store=True, readonly=False, precompute=True,
-        domain="[('id', 'child_of', parent.location_dest_id), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
-    is_done = fields.Boolean('Done', compute='_compute_is_done', inverse='_set_is_done')
-    state = fields.Selection([
-        ('draft', 'Draft'),
-        ('confirmed', 'Confirmed'),
-        ('assigned', 'Reserved'),
-        ('new', 'New'),
-        ('done', 'Done'),
-        ('cancel', 'Cancelled'),
-    ],string='State', compute='_compute_state')
-    is_fresh_package = fields.Boolean(compute='_compute_fresh_pack')
-
-    picking_type_code = fields.Selection(related='picking_id.picking_type_code')
-    show_lots_m2o = fields.Boolean(compute='_compute_show_lot')
-    show_lots_text = fields.Boolean(compute='_compute_show_lot')
-    company_id = fields.Many2one('res.company', 'Company', required=True, index=True)
-
-    @api.depends('move_line_ids', 'move_line_ids.quantity')
-    def _compute_is_done(self):
-        for package_level in self:
-            # If it is an existing package
-            if package_level.is_fresh_package:
-                package_level.is_done = True
-            else:
-                package_level.is_done = package_level._check_move_lines_map_quant_package(package_level.package_id, only_picked=True)
-
-    def _set_is_done(self):
-        for package_level in self:
-            if package_level.is_done:
-                if not package_level.is_fresh_package:
-                    ml_update_dict = defaultdict(float)
-                    package_level.picking_id.move_line_ids.filtered(
-                        lambda ml: not ml.package_level_id and ml.package_id == package_level.package_id
-                    ).unlink()
-                    for quant in package_level.package_id.quant_ids:
-                        corresponding_mls = package_level.move_line_ids.filtered(lambda ml: ml.product_id == quant.product_id and ml.lot_id == quant.lot_id)
-                        to_dispatch = quant.quantity
-                        if corresponding_mls:
-                            for ml in corresponding_mls:
-                                qty = min(to_dispatch, ml.move_id.product_qty) if len(corresponding_mls) > 1 else to_dispatch
-                                to_dispatch = to_dispatch - qty
-                                ml_update_dict[ml] += qty
-                                if ml.product_id.uom_id.is_zero(to_dispatch):
-                                    break
-                        else:
-                            corresponding_move = package_level.move_ids.filtered(lambda m: m.product_id == quant.product_id)[:1]
-                            self.env['stock.move.line'].create({
-                                'location_id': package_level.location_id.id,
-                                'location_dest_id': package_level.location_dest_id.id,
-                                'picking_id': package_level.picking_id.id,
-                                'product_id': quant.product_id.id,
-                                'quantity': quant.quantity,
-                                'product_uom_id': quant.product_id.uom_id.id,
-                                'lot_id': quant.lot_id.id,
-                                'package_id': package_level.package_id.id,
-                                'result_package_id': package_level.package_id.id,
-                                'package_level_id': package_level.id,
-                                'move_id': corresponding_move.id,
-                                'owner_id': quant.owner_id.id,
-                                'picked': True,
-                            })
-                    for rec, quant in ml_update_dict.items():
-                        rec.quantity = quant
-                        rec.picked = True
-            else:
-                package_level.move_line_ids.unlink()
-
-    @api.depends('move_line_ids', 'move_line_ids.package_id', 'move_line_ids.result_package_id')
-    def _compute_fresh_pack(self):
-        for package_level in self:
-            if not package_level.move_line_ids or all(ml.package_id and ml.package_id == ml.result_package_id for ml in package_level.move_line_ids):
-                package_level.is_fresh_package = False
-            else:
-                package_level.is_fresh_package = True
-
-    @api.depends('move_ids', 'move_ids.state', 'move_line_ids', 'move_line_ids.state')
-    def _compute_state(self):
-        for package_level in self:
-            if not package_level.move_ids and not package_level.move_line_ids:
-                package_level.state = 'draft'
-            elif not package_level.move_line_ids and package_level.move_ids.filtered(lambda m: m.state not in ('done', 'cancel')):
-                package_level.state = 'confirmed'
-            elif package_level.move_line_ids and not package_level.move_line_ids.filtered(lambda ml: ml.state in ('done', 'cancel')):
-                if package_level.is_fresh_package:
-                    package_level.state = 'new'
-                elif package_level._check_move_lines_map_quant_package(package_level.package_id):
-                    package_level.state = 'assigned'
-                else:
-                    package_level.state = 'confirmed'
-            elif package_level.move_line_ids.filtered(lambda ml: ml.state =='done'):
-                package_level.state = 'done'
-            elif package_level.move_line_ids.filtered(lambda ml: ml.state == 'cancel') or package_level.move_ids.filtered(lambda m: m.state == 'cancel'):
-                package_level.state = 'cancel'
-            else:
-                package_level.state = 'draft'
-
-    def _compute_show_lot(self):
-        for package_level in self:
-            if any(ml.product_id.tracking != 'none' for ml in package_level.move_line_ids):
-                if package_level.picking_id.picking_type_id.use_existing_lots or package_level.state == 'done':
-                    package_level.show_lots_m2o = True
-                    package_level.show_lots_text = False
-                else:
-                    if self.picking_id.picking_type_id.use_create_lots and package_level.state != 'done':
-                        package_level.show_lots_m2o = False
-                        package_level.show_lots_text = True
-                    else:
-                        package_level.show_lots_m2o = False
-                        package_level.show_lots_text = False
-            else:
-                package_level.show_lots_m2o = False
-                package_level.show_lots_text = False
-
-    def _generate_moves(self):
-        for package_level in self:
-            if package_level.package_id:
-                for quant in package_level.package_id.quant_ids:
-                    self.env['stock.move'].create({
-                        'picking_id': package_level.picking_id.id,
-                        'product_id': quant.product_id.id,
-                        'product_uom_qty': quant.quantity,
-                        'product_uom': quant.product_id.uom_id.id,
-                        'location_id': package_level.location_id.id,
-                        'location_dest_id': package_level.location_dest_id.id,
-                        'package_level_id': package_level.id,
-                        'company_id': package_level.company_id.id,
-                    })
-
-    @api.model_create_multi
-    def create(self, vals_list):
-        package_levels = super().create(vals_list)
-        for package_level, vals in zip(package_levels, vals_list):
-            if vals.get('location_dest_id') and not self.env.context.get('from_put_in_pack'):
-                package_level.move_line_ids.write({'location_dest_id': vals['location_dest_id']})
-                package_level.move_ids.write({'location_dest_id': vals['location_dest_id']})
-        return package_levels
-
-    def write(self, vals):
-        result = super().write(vals)
-        if vals.get('location_dest_id'):
-            self.mapped('move_line_ids').write({'location_dest_id': vals['location_dest_id']})
-            self.mapped('move_ids').write({'location_dest_id': vals['location_dest_id']})
-        return result
-
-    def unlink(self):
-        self.mapped('move_ids').write({'package_level_id': False})
-        self.mapped('move_line_ids').write({'result_package_id': False})
-        return super().unlink()
-
-    def _check_move_lines_map_quant_package(self, package, only_picked=False):
-        mls = self.move_line_ids
-        if only_picked:
-            mls = mls.filtered(lambda ml: ml.picked)
-        return package._check_move_lines_map_quant(mls)
-
-    @api.depends('package_id', 'state', 'is_fresh_package', 'move_ids', 'move_line_ids')
-    def _compute_location_id(self):
-        for pl in self:
-            if pl.state == 'new' or pl.is_fresh_package:
-                pl.location_id = False
-            elif pl.state != 'done' and pl.package_id:
-                pl.location_id = pl.package_id.location_id
-            elif pl.state == 'confirmed' and pl.move_ids:
-                pl.location_id = pl.move_ids[0].location_id
-            elif pl.state in ('assigned', 'done') and pl.move_line_ids:
-                pl.location_id = pl.move_line_ids[0].location_id
-            else:
-                pl.location_id = pl.picking_id.location_id
-
-    @api.depends('picking_id', 'picking_id.location_dest_id')
-    def _compute_location_dest_id(self):
-        for pl in self:
-            pl.location_dest_id = pl.picking_id.location_dest_id
-
-    def action_show_package_details(self):
-        self.ensure_one()
-        view = self.env.ref('stock.package_level_form_edit_view')
-
-        return {
-            'name': _('Package Content'),
-            'type': 'ir.actions.act_window',
-            'view_mode': 'form',
-            'res_model': 'stock.package_level',
-            'views': [(view.id, 'form')],
-            'view_id': view.id,
-            'target': 'new',
-            'res_id': self.id,
-        }

--- a/addons/stock/models/stock_picking.py
+++ b/addons/stock/models/stock_picking.py
@@ -615,8 +615,6 @@ class StockPicking(models.Model):
         compute="_compute_location_id", store=True, precompute=True, readonly=False,
         check_company=True, required=True)
     move_ids = fields.One2many('stock.move', 'picking_id', string="Stock Moves", copy=True)
-    move_ids_without_package = fields.One2many(
-        'stock.move', 'picking_id', string="Stock move", domain=['|', ('package_level_id', '=', False), ('picking_type_entire_packs', '=', False)])
     has_scrap_move = fields.Boolean(
         'Has Scrap Moves', compute='_has_scrap_move')
     picking_type_id = fields.Many2one(
@@ -642,7 +640,6 @@ class StockPicking(models.Model):
         default=lambda self: self.env.user, copy=False
     )
     move_line_ids = fields.One2many('stock.move.line', 'picking_id', 'Operations')
-    move_line_ids_without_package = fields.One2many('stock.move.line', 'picking_id', 'Operations without package', domain=['|',('package_level_id', '=', False), ('picking_type_entire_packs', '=', False)])
     packages_count = fields.Integer('Packages Count', compute='_compute_packages_count')
     show_check_availability = fields.Boolean(
         compute='_compute_show_check_availability',
@@ -678,8 +675,6 @@ class StockPicking(models.Model):
     show_operations = fields.Boolean(related='picking_type_id.show_operations')
     show_lots_text = fields.Boolean(compute='_compute_show_lots_text')
     has_tracking = fields.Boolean(compute='_compute_has_tracking')
-    package_level_ids = fields.One2many('stock.package_level', 'picking_id')
-    package_level_ids_details = fields.One2many('stock.package_level', 'picking_id')
     products_availability = fields.Char(
         string="Product Availability", compute='_compute_products_availability',
         help="Latest product availability status of the picking")
@@ -1069,26 +1064,14 @@ class StockPicking(models.Model):
     def _onchange_picking_type(self):
         if self.picking_type_id and self.state == 'draft':
             self = self.with_company(self.company_id)
-            # The compute store doesn't work in case of One2many inverse (move_ids_without_package)
-            (self.move_ids | self.move_ids_without_package).filtered(
+            self.move_ids.filtered(
                 lambda m: m.picking_type_id != self.picking_type_id
             ).picking_type_id = self.picking_type_id
-            (self.move_ids | self.move_ids_without_package).company_id = self.company_id
-
-    @api.onchange('location_dest_id')
-    def _onchange_location_dest_id(self):
-        moves = self.move_ids_without_package
-        if any(not move._origin for move in moves):
-            # Because of an ORM limitation, the new SM defined in self.move_ids_without_package are not set in
-            # self.move_ids. Since the user edits the destination location, the ORM will check which SM must be
-            # recomputed (cf dependencies of SM._compute_location_dest_id). But, to do so, the ORM will look at
-            # self.move_ids, i.e.: it will not call the compute method for the new SM. We therefore have to
-            # manually trigger the compute method
-            self.env.add_to_compute(moves._fields['location_dest_id'], moves)
+            self.move_ids.company_id = self.company_id
 
     @api.onchange('location_id')
     def _onchange_location_id(self):
-        (self.move_ids | self.move_ids_without_package).location_id =  self.location_id
+        self.move_ids.location_id = self.location_id
         for move in self.move_ids.filtered(lambda m: m.move_orig_ids):
             for ml in move.move_line_ids:
                 parent_path = [int(loc_id) for loc_id in ml.location_id.parent_path.split('/')[:-1]]
@@ -1152,7 +1135,7 @@ class StockPicking(models.Model):
             after_vals['partner_id'] = vals['partner_id']
         if after_vals:
             self.move_ids.filtered(lambda move: not move.scrapped).write(after_vals)
-        if vals.get('move_ids') or vals.get('move_ids_without_package'):
+        if vals.get('move_ids'):
             self._autoconfirm_picking()
 
         return res
@@ -1176,7 +1159,6 @@ class StockPicking(models.Model):
 
     def action_confirm(self):
         self._check_company()
-        self.mapped('package_level_ids').filtered(lambda pl: pl.state == 'draft' and not pl.move_ids)._generate_moves()
         # call `_action_confirm` on every draft move
         self.move_ids.filtered(lambda move: move.state == 'draft')._action_confirm()
 
@@ -1190,7 +1172,6 @@ class StockPicking(models.Model):
         also impact the state of the picking as it is computed based on move's states.
         @return: True
         """
-        self.mapped('package_level_ids').filtered(lambda pl: pl.state == 'draft' and not pl.move_ids)._generate_moves()
         self.filtered(lambda picking: picking.state == 'draft').action_confirm()
         moves = self.move_ids.filtered(lambda move: move.state not in ('draft', 'cancel', 'done')).sorted(
             key=lambda move: (-int(move.priority), not bool(move.date_deadline), move.date_deadline, move.date, move.id)
@@ -1381,7 +1362,6 @@ class StockPicking(models.Model):
 
     def do_unreserve(self):
         self.move_ids._do_unreserve()
-        self.package_level_ids.filtered(lambda p: not p.move_ids).unlink()
 
     def button_validate(self):
         self = self.filtered(lambda p: p.state != 'done')
@@ -1536,7 +1516,7 @@ class StockPicking(models.Model):
         for picking in self:
             if picking.state in ('done', 'cancel'):
                 continue
-            if not picking.move_ids and not picking.package_level_ids:
+            if not picking.move_ids:
                 continue
             if any(move.additional for move in picking.move_ids):
                 picking.action_confirm()
@@ -1571,7 +1551,6 @@ class StockPicking(models.Model):
             if moves_to_backorder:
                 backorder_picking = picking._create_backorder_picking()
                 moves_to_backorder.write({'picking_id': backorder_picking.id, 'picked': False})
-                moves_to_backorder.move_line_ids.package_level_id.write({'picking_id': backorder_picking.id})
                 moves_to_backorder.mapped('move_line_ids').write({'picking_id': backorder_picking.id})
                 backorders |= backorder_picking
                 backorder_picking.user_id = False
